service/history/historyEngine.go

```diff
@@ -2666,40 +2672,21 @@ func (e *historyEngineImpl) NotifyNewHistoryEvent(
 	e.eventNotifier.NotifyNewHistoryEvent(notification)
 }
 
-func (e *historyEngineImpl) NotifyNewTransferTasks(
-	clusterName string,
-	tasks []tasks.Task,
-) {
-	if len(tasks) > 0 {
-		e.txProcessor.NotifyNewTask(clusterName, tasks)
-	}
-}
-
-func (e *historyEngineImpl) NotifyNewTimerTasks(
+func (e *historyEngineImpl) NotifyNewTasks(
 	clusterName string,
-	tasks []tasks.Task,
-) {
-
-	if len(tasks) > 0 {
-		e.timerProcessor.NotifyNewTimers(clusterName, tasks)
-	}
-}
-
-func (e *historyEngineImpl) NotifyNewReplicationTasks(
-	tasks []tasks.Task,
-) {
-
-	if len(tasks) > 0 && e.replicatorProcessor != nil {
-		e.replicatorProcessor.NotifyNewTasks(tasks)
-	}
-}
-
-func (e *historyEngineImpl) NotifyNewVisibilityTasks(
-	tasks []tasks.Task,
+	newTasks map[tasks.Category][]tasks.Task,
 ) {
+	for category, tasksByCategory := range newTasks {
+		if category == tasks.CategoryReplication {
```

Comment:
this is ugly

Reply:
Yeah let me add a TODO here to get rid of this special case.

I will probably still have replication queue processor implement the `queues.Processor` interface but leave most method implementation empty. Then in history engine, the replication queue processor can be added to the queue processor list (thus removing the special case here). 
We still need another reference in history engine to the more specific replication queue processor type/interface so we can call various methods on it from history engine (or we do type assertion, but I think that's ugly as well...).

we can continue the this discussion here or offline after the pr is merged.